### ISSUE TITLE
Dtensor shard uses more gpu memory than raw tensor

### ISSUE THREAD

**Issue by v4if** [pytorch/pytorch#133549]
### 🐛 Describe the bug

Dtensor shard uses more gpu memory than raw tensor. 
With test, Shard gpu mem: 21890MiB > Replicate gpu mem: 17448MiB > Raw tensor gpu mem: 16804MiB.
Confused for a long time.

```python
# torchrun --nproc_per_node=4 test_dtensor.py

import os
import torch
import torch.distributed as dist
from torch.distributed._tensor import DTensor, Shard, Replicate, distribute_tensor, distribute_module, init_device_mesh

mesh = init_device_mesh("cuda", (int(os.environ["WORLD_SIZE"]),))

def test_raw_tensor():
    big_tensor = torch.randn((4, 1024*1024*1024)).to("cuda")

def test_shard():
    big_tensor = torch.randn((4, 1024*1024*1024))
    my_dtensor = distribute_tensor(big_tensor, mesh, [Shard(dim=0)])

def test_replicate():
    big_tensor = torch.randn((4, 1024*1024*1024))
    my_dtensor = distribute_tensor(big_tensor, mesh, [Replicate()])

# test_raw_tensor()
test_shard()
# test_replicate()

print("wait for gpu inspect...")
import time
time.sleep(60)

```

### Versions

```text
PyTorch version: 2.4.0+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.31

Python version: 3.10.14 (main, May  6 2024, 19:42:50) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.10.134-13.al8.x86_64-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 12.1.105
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA A800-SXM4-80GB
GPU 1: NVIDIA A800-SXM4-80GB
GPU 2: NVIDIA A800-SXM4-80GB
GPU 3: NVIDIA A800-SXM4-80GB
GPU 4: NVIDIA A800-SXM4-80GB
GPU 5: NVIDIA A800-SXM4-80GB
GPU 6: NVIDIA A800-SXM4-80GB
GPU 7: NVIDIA A800-SXM4-80GB

Nvidia driver version: 535.161.08
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   46 bits physical, 57 bits virtual
CPU(s):                          128
On-line CPU(s) list:             0-123
Off-line CPU(s) list:            124-127
Thread(s) per core:              1
Core(s) per socket:              32
Socket(s):                       2
NUMA node(s):                    2
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           106
Model name:                      Intel(R) Xeon(R) Platinum 8369B CPU @ 2.90GHz
Stepping:                        6
CPU MHz:                         3499.994
CPU max MHz:                     3500.0000
CPU min MHz:                     800.0000
BogoMIPS:                        5800.00
Virtualization:                  VT-x
L1d cache:                       1.5 MiB
L1i cache:                       1 MiB
L2 cache:                        40 MiB
L3 cache:                        48 MiB
NUMA node0 CPU(s):               0-31,64-95
NUMA node1 CPU(s):               32-63,96-127
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect wbnoinvd dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq rdpid fsrm md_clear pconfig flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] torch==2.4.0
[pip3] torchvision==0.19.0
[pip3] triton==3.0.0
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] torch                     2.4.0                    pypi_0    pypi
[conda] torchvision               0.19.0                   pypi_0    pypi
[conda] triton                    3.0.0                    pypi_0    pypi
```

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @tianyu-l

**Comment by weifengpy** [comment]
this is good question! I dumped memory snapshot with following code

for `test_shard`, the peak memory happens at `distribute_tensor(big_tensor, mesh, [Shard(dim=0)])`, where we have 16GB + 16GB / 4 = 21 GB. When finishing `test_shard`, pytorch active memory is reduced to 4 GB. However, pytorch has [Cuda Caching Allocator](https://zdevito.github.io/2022/08/04/cuda-caching-allocator.html). The memory is not truly released (nvidia-smi still shows 21 GB+). The memory is kept in a pool so any pytorch tensors could grab it again. For training loop, keeping a pool of memory works well and it avoid asking for cuda memory allocation frequently

<img width="1477" alt="Screenshot 2024-08-18 at 11 07 40 PM" src="https://github.com/user-attachments/assets/f2ab6682-d2d3-45c1-a13c-d84ae7f2e156">

I do not see significant difference between replicate and raw tensor though. they both look like this

<img width="699" alt="Screenshot 2024-08-18 at 11 15 20 PM" src="https://github.com/user-attachments/assets/cc941b7b-bbf3-4325-9967-e82643818661">


```
# torchrun --nproc_per_node=4 test_dtensor.py

import os
import torch
import torch.distributed as dist
from torch.distributed._tensor import DTensor, Shard, Replicate, distribute_tensor, distribute_module, init_device_mesh
import pickle
torch.cuda.memory._record_memory_history(max_entries=100000)

mesh = init_device_mesh("cuda", (int(os.environ["WORLD_SIZE"]),))

def test_raw_tensor():
    big_tensor = torch.randn((4, 1024*1024*1024)).to("cuda")

def test_shard():
    big_tensor = torch.randn((4, 1024*1024*1024))
    my_dtensor = distribute_tensor(big_tensor, mesh, [Shard(dim=0)])

def test_replicate():
    big_tensor = torch.randn((4, 1024*1024*1024))
    my_dtensor = distribute_tensor(big_tensor, mesh, [Replicate()])

#test_raw_tensor()
test_shard()
#test_replicate()
snapshot = torch.cuda.memory._snapshot()
if torch.distributed.get_rank() == 0:
    pickle.dump(snapshot, open("test_dtensor.pickle", 'wb'))
```

**Comment by weifengpy** [comment]
(continued) for test_shard, right now we move `big_tensor` first (orange) to `cuda` and do NCCL scatter (red). It's possible to implement it as split `big_tensor` on cpu first and move the 1/N to `cuda`

curious what's is the real use case? do you get the big_tensor from model init? For DTensor, usually we do meta init first (`model.parameters()` is chunked into 1/N), then we do DTensor init (example code)[https://github.com/pytorch/pytorch/blob/main/test/distributed/_tensor/test_random_ops.py#L331-L332]

**Comment by weifengpy** [comment]
if you are using `nvidia-smi` to check memory usage, calling `torch.cuda.empty_cache()` will truly release the memory and you should only see 1/4 memory (red color) at the end

**Comment by wanchaol** [comment]
> (continued) for test_shard, right now we move `big_tensor` first (orange) to `cuda` and do NCCL scatter (red). It's possible to implement it as split `big_tensor` on cpu first and move the 1/N to `cuda`. But curious to understand your use case. Do you get the big_tensor from model init? For DTensor, usually we do meta init first (`model.parameters()` is chunked into 1/N), then we do DTensor init (example code)[https://github.com/pytorch/pytorch/blob/main/test/distributed/_tensor/test_random_ops.py#L331-L332]

Yeah this is exactly right! To add more more color, @v4if We move big tensor to cuda first then perform scatter. The reason why we do this instead of CPU split first then move it to GPU is because we need to preserve the single device semantic, where CPU split first would not preserve single device semantic as every rank have a different RNG state. This means we need to perform the scatter (for sharding) using one rank's big/full_tensor (we choose to use rank 0 on the device mesh dim), since it's a cuda device mesh, we perform scatter using NCCL, that's why we need to bring the full tensor to GPU first. 

If you want no initial memory spike (even though the memory released right after but still kept in the cuda caching allocator for later tensor usage), you can use meta tensor init as @weifengpy suggested

**Comment by v4if** [comment]
> > (continued) for test_shard, right now we move `big_tensor` first (orange) to `cuda` and do NCCL scatter (red). It's possible to implement it as split `big_tensor` on cpu first and move the 1/N to `cuda`. But curious to understand your use case. Do you get the big_tensor from model init? For DTensor, usually we do meta init first (`model.parameters()` is chunked into 1/N), then we do DTensor init (example code)[https://github.com/pytorch/pytorch/blob/main/test/distributed/_tensor/test_random_ops.py#L331-L332]
> 
> Yeah this is exactly right! To add more more color, @v4if We move big tensor to cuda first then perform scatter. The reason why we do this instead of CPU split first then move it to GPU is because we need to preserve the single device semantic, where CPU split first would not preserve single device semantic as every rank have a different RNG state. This means we need to perform the scatter (for sharding) using one rank's big/full_tensor (we choose to use rank 0 on the device mesh dim), since it's a cuda device mesh, we perform scatter using NCCL, that's why we need to bring the full tensor to GPU first.
> 
> If you want no initial memory spike (even though the memory released right after but still kept in the cuda caching allocator for later tensor usage), you can use meta tensor init as @weifengpy suggested

Calling torch.cuda.empty_cache() frees cached memory, memory usage dropped to 1/4 of the original tensor.
But move big tensor to cuda seems to occur on all ranks, not just rank0. When using nvidia-smi to observe, the memory memory on each rank before empty_cache is 21GiB. The semantics here are not quite like scatter, but more like broadcast.

**Comment by wconstab** [comment]
I am closing this issue as not a bug.  The behavior is as designed, and there are workarounds for reducing the memory usage.  However please feel free to file a feature-request issue if you want to ask for a different behavior.
